### ISSUE TITLE
vault-token arg passed with "nomad job run" doesn't make it to task

### ISSUE THREAD

**Issue by nathanhruby** [hashicorp/nomad#8040]
### Nomad version
Nomad v0.11.2

### Operating system and Environment details
- Ubuntu 18.04
- Current version of nomad, vault, and consul OSS

### Issue
`nomad run -vault-token="token" path/job.nomad` doesn't seem to get the vault token to the client machine, so templates using vault fail.  

It doesn't seem like this is the correct behavior, but maybe I am misinterpreting the documentation?  As I read it, it seems like a passed token would be handed down to the client with the job.  It's unclear if it'd be *renewed* it it seems like it should *work*.

### Reproduction steps
- vault server config vault{} with:
```hcl
vault {
  allow_unauthenticated = false
  address = "https://active.vault.service.consul:8200"
}
```
- vault client configure vault{} with
```hcl
vault {
  enabled = true
  address = "https://active.vault.service.consul:8200"
}
```
- vault token create -policy mypol
- nomad job run -vault-token=TOKEN path/job.nomad

#### Expected
Template renders, job runs, vault keeps passed token renewed

#### Actual 
Job sits in pending on the client waiting vault read

### Job file (if appropriate)
```hcl
job "fabio" {
  datacenters = ["thisdc"]
  type = "system"

  group "web" {
    task "lb" {
      driver = "docker"
      config {
        image = "fabiolb/fabio"
        network_mode = "host"
        cap_add = [ "NET_BIND_SERVICE" ]
      }
      env {
        proxy_addr = ":80"
      }
      template {
        destination = "local/file.yml"
        env = true
        data = <<EOH
{{ with secret "consul/creds/fabio" }}
CONSUL_HTTP_TOKEN="{{ .Data.token }}"
{{ end }}
EOH

      }
      resources {
        cpu    = 101
        memory = 64
        network {
          mbits = 10
          port "lb_http" {
            static = 80
          }
          port "ui" {
            static = 9998
          }
        }
      }
    }
  }
}

```
### Nomad Client logs (if appropriate)
```
May 23 00:11:38 nomad-client-i-0daa6f8a33ea83dd5 nomad[18745]:     2020/05/23 00:11:38.202855 [INFO] (runner) creating new runner (dry: false, once: false)
May 23 00:11:38 nomad-client-i-0daa6f8a33ea83dd5 nomad[18745]:     2020/05/23 00:11:38.202986 [INFO] (runner) creating watcher
May 23 00:11:38 nomad-client-i-0daa6f8a33ea83dd5 nomad[18745]:     2020/05/23 00:11:38.203080 [INFO] (runner) starting
May 23 00:11:42 nomad-client-i-0daa6f8a33ea83dd5 nomad[18745]:     2020/05/23 00:11:42.047044 [WARN] (view) vault.read(consul/creds/fabio): vault.read(consul/creds/fabio): Error making API request.
May 23 00:11:42 nomad-client-i-0daa6f8a33ea83dd5 nomad[18745]: URL: GET https://active.vault.service.consul:8200/v1/consul/creds/fabio
May 23 00:11:42 nomad-client-i-0daa6f8a33ea83dd5 nomad[18745]: Code: 400. Errors:
May 23 00:11:42 nomad-client-i-0daa6f8a33ea83dd5 nomad[18745]: * missing client token (retry attempt 1 after "250ms")
May 23 00:11:42 nomad-client-i-0daa6f8a33ea83dd5 nomad[18745]:     2020/05/23 00:11:42.298463 [WARN] (view) vault.read(consul/creds/fabio): vault.read(consul/creds/fabio): Error making API request.
May 23 00:11:42 nomad-client-i-0daa6f8a33ea83dd5 nomad[18745]: URL: GET https://active.vault.service.consul:8200/v1/consul/creds/fabio
May 23 00:11:42 nomad-client-i-0daa6f8a33ea83dd5 nomad[18745]: Code: 400. Errors:
May 23 00:11:42 nomad-client-i-0daa6f8a33ea83dd5 nomad[18745]: * missing client token (retry attempt 2 after "500ms")
May 23 00:11:42 nomad-client-i-0daa6f8a33ea83dd5 nomad[18745]:     2020/05/23 00:11:42.799927 [WARN] (view) vault.read(consul/creds/fabio): vault.read(consul/creds/fabio): Error making API request.
May 23 00:11:42 nomad-client-i-0daa6f8a33ea83dd5 nomad[18745]: URL: GET https://active.vault.service.consul:8200/v1/consul/creds/fabio
May 23 00:11:42 nomad-client-i-0daa6f8a33ea83dd5 nomad[18745]: Code: 400. Errors:
May 23 00:11:42 nomad-client-i-0daa6f8a33ea83dd5 nomad[18745]: * missing client token (retry attempt 3 after "1s")
May 23 00:11:43 nomad-client-i-0daa6f8a33ea83dd5 nomad[18745]:     2020/05/23 00:11:43.801419 [WARN] (view) vault.read(consul/creds/fabio): vault.read(consul/creds/fabio): Error making API request.
```

**Comment by leungster** [comment]
Your job file is missing a [vault stanza](https://www.nomadproject.io/docs/job-specification/vault)

The documentation for the CLI is confusing because it sounds like it's passing the token to the job when it executes. However, it's really storing it in the job run API request and the server uses that token to validate if the requesting user has permissions to the vault policies defined in the job.

>  -vault-token
    If set, the passed Vault token is stored in the job before sending to the
    Nomad servers. This allows passing the Vault token without storing it in
    the job file. This overrides the token found in $VAULT_TOKEN environment
    variable and that found in the job.

The vault token is cleared after validating it.
https://github.com/hashicorp/nomad/blob/a9dd409207e6938526f62f23111402d20856940e/nomad/job_endpoint.go#L235-L236

**Comment by nathanhruby** [comment]
Yeah, you answered my question:

> The documentation for the CLI is confusing because it sounds like ...

Implicit was that if one passed a token at submit time, one wouldn't need a vault stanza in the job because the requisite work was already done.

I think this is probably a doc bug to make the validation only nature clearer.  IMO, the integration docs here are a bit hand wavy, so it's easy to get led astray.

**Comment by tgross** [comment]
Thanks @leungster and sorry about the delay on circling back to this @nathanhruby. I ran into similar trouble recently while writing some end-to-end tests, so I feel your pain on the docs here. I'll mark this as something to improve in the documentation.
